[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/CO-CONNECT/co-connect-tools/HEAD)


## Installing

The best way is to install the module via `pip`. 

In [ ]:
!pip3 install co-connect-tools -q

In [1]:
import coconnect.tools
data = coconnect.tools.load_json('example/sample_config/panther_structural_mapping.json')
coconnect.tools.extract.make_class(data,'Panther')

Recreating file /Users/calummacdonald/Usher/CO-CONNECT/Software/docs/docs/co-connect-tools/coconnect/cdm/classes/Panther.py


In [2]:
print (coconnect.tools.get_classes(format=True))

{
      "Panther": {
            "module": "coconnect.cdm.classes.Panther",
            "path": "/Users/calummacdonald/Usher/CO-CONNECT/Software/docs/docs/co-connect-tools/coconnect/cdm/classes/Panther.py",
            "last-modified": "2021-03-31 11:58:08"
      }
}


In [3]:
f_map = coconnect.tools.get_file_map_from_dir('example/sample_input_data/')
f_map

{'tracker.csv': '/Users/calummacdonald/Usher/CO-CONNECT/Software/docs/docs/co-connect-tools/coconnect/data/example/sample_input_data/tracker.csv',
 'demographic.csv': '/Users/calummacdonald/Usher/CO-CONNECT/Software/docs/docs/co-connect-tools/coconnect/data/example/sample_input_data/demographic.csv',
 'questionnaire.csv': '/Users/calummacdonald/Usher/CO-CONNECT/Software/docs/docs/co-connect-tools/coconnect/data/example/sample_input_data/questionnaire.csv'}

In [4]:
inputs = coconnect.tools.load_csv(f_map)

In [5]:
from coconnect.cdm.classes.Panther import Panther
panther = Panther(inputs=inputs)

2021-03-31 11:58:09 - Panther - INFO - CommonDataModel created
2021-03-31 11:58:09 - Panther - WARNING - trying to set index 'id number' for 'classifedroledata.csv' but this has not been loaded as an inputs!
2021-03-31 11:58:09 - Panther - WARNING - trying to set index 'serum 7patient id' for 'freezer.csv' but this has not been loaded as an inputs!
2021-03-31 11:58:09 - Panther - WARNING - trying to set index 'sample id' for 'hta log 04062020.csv' but this has not been loaded as an inputs!
2021-03-31 11:58:09 - Panther - WARNING - trying to set index 'patient id' for 'serum 1.csv' but this has not been loaded as an inputs!
2021-03-31 11:58:09 - Panther - WARNING - trying to set index 'patient id' for 'serum short.csv' but this has not been loaded as an inputs!
2021-03-31 11:58:09 - Panther - WARNING - trying to set index 'id' for 'visit 1.csv' but this has not been loaded as an inputs!


In [6]:
panther.process()

2021-03-31 11:58:09 - Panther - INFO - for person: found 1 object
2021-03-31 11:58:09 - Panther - INFO - working on <class 'coconnect.cdm.objects.person.Person'>
2021-03-31 11:58:10 - person_0 - WARNING - person_id 7680/10000 indicies (person_id) are duplicated
2021-03-31 11:58:10 - person_0 - WARNING -       if this is synthetic data... dont worry about it
2021-03-31 11:58:10 - person_0 - WARNING - gender_concept_id 9367/10000 indicies (person_id) are duplicated
2021-03-31 11:58:10 - person_0 - WARNING -       if this is synthetic data... dont worry about it
2021-03-31 11:58:10 - person_0 - WARNING - year_of_birth 9371/10000 indicies (person_id) are duplicated
2021-03-31 11:58:10 - person_0 - WARNING -       if this is synthetic data... dont worry about it
2021-03-31 11:58:10 - person_0 - WARNING - month_of_birth 9371/10000 indicies (person_id) are duplicated
2021-03-31 11:58:10 - person_0 - WARNING -       if this is synthetic data... dont worry about it
2021-03-31 11:58:10 - person_

In [7]:
panther.omop['person']

gender_concept_id  year_of_birth  month_of_birth  day_of_birth  \
person_id                                                                   
1                       <NA>           <NA>            <NA>          <NA>   
2                       <NA>           <NA>            <NA>          <NA>   
3                       <NA>           <NA>            <NA>          <NA>   
4                       <NA>           <NA>            <NA>          <NA>   
5                       <NA>           <NA>            <NA>          <NA>   
...                      ...            ...             ...           ...   
2321                    8532           1973               1             5   
2321                    8507           1993               3             9   
2321                    8532           1967               3            28   
2321                    8507           1959               7            19   
2321                    8532           1976              10            11   

                birth_datetime  race_concept_id  ethnicity_concept_id  \
person_id                                                               
1                          NaN             <NA>                  <NA>   
2                          NaN             <NA>                  <NA>   
3                          NaN             <NA>                  <NA>   
4                          NaN             <NA>                  <NA>   
5                          NaN             <NA>                  <NA>   
...                        ...              ...                   ...   
2321       1973-01-05 00:00:00          4196428                  <NA>   
2321       1993-03-09 00:00:00          4185920                  <NA>   
2321       1967-03-28 00:00:00          4196428                  <NA>   
2321       1959-07-19 00:00:00          4196428                  <NA>   
2321       1976-10-11 00:00:00          4196428                  <NA>   

           location_id  provider_id  care_site_id person_source_value  \
person_id                                                               
1                 <NA>         <NA>          <NA>                       
2                 <NA>         <NA>          <NA>                       
3                 <NA>         <NA>          <NA>                       
4                 <NA>         <NA>          <NA>                       
5                 <NA>         <NA>          <NA>                       
...                ...          ...           ...                 ...   
2321           4021920         <NA>          <NA>                       
2321              <NA>         <NA>          <NA>                       
2321           4021920         <NA>          <NA>                       
2321           4021920         <NA>          <NA>                       
2321           4021920         <NA>          <NA>                       

          gender_source_value  gender_source_concept_id race_source_value  \
person_id                                                                   
1                                                  <NA>                     
2                                                  <NA>                     
3                                                  <NA>                     
4                                                  <NA>                     
5                                                  <NA>                     
...                       ...                       ...               ...   
2321                        F                      8532     WHITE BRITISH   
2321                        M                      8507            INDIAN   
2321                        F                      8532     WHITE BRITISH   
2321                        M                      8507     WHITE BRITISH   
2321                        F                      8532     WHITE BRITISH   

           race_source_concept_id ethnicity_source_value  \
person_id                                      

## Load Inputs

To run the tool you need to load some input datasets, and specify how to map the fields 

The data will be loaded into pandas dataframes that we'll use for some visualisations of what the input `csv` files will look like

### Source data

This data is synthetic data that has been produced by [OHDSI](http://ohdsi.org/) which simply details a record of patients.

_Note: that these example data files will be stored in `<install_dir>/lib/python3.8/site-packages/coconnect/`, a directory that `ETLTool` will be looking in. For your own files, you should specify the full path to the inputs_

In [ ]:
f_input_data = 'sample_input_data/patients_sample.csv'
etltool.load_input_data([f_input_data])

Verify what files have been loaded, by default the input dataset is mapped to to a name via `/path/<name>.csv`

In [ ]:
etltool.get_input_names()

Sample (3 entries) what this input data looks like. __Note__ becareful using this method with a large dataset

In [ ]:
df_input = etltool.get_input_df('patients_sample.csv')
df_input.sample(3)

### Structural Mapping

Next we use another `csv` file to define how to map different fields in the source data to a [Common Data Model (CDM)](https://www.ohdsi.org/data-standardization/the-common-data-model/).

In this example, the CDM that the source data (`patients_sample`)  is being mappped to is the `person` CDM.

There are three rules defined:
1. Performs a straight one-to-one mapping between the field `id` in the source field to the `person_id` field of the `person` CDM
2. Performs a mapping with the operation/function `extract year` 
3. Performs a term mapping which is defined in the term mapping `csv` file, see the next section for more information 


In [ ]:
f_structural_mapping = 'sample_input_data/rules1.csv'
etltool.load_structural_mapping(f_structural_mapping) 
etltool.get_structural_mapping_df()

#### Testing operations
The 2nd rule defined the operation `extract year`, this is a default operation defined in `etltool`, here is a quick example of how it works..

Load the function

In [ ]:
fn_extract_year = etltool.allowed_operations['EXTRACT_YEAR']
fn_extract_year

For example, taking the `BIRTHDATE` columns, which looks like:

In [ ]:
df_input['BIRTHDATE'].head(4)

The function can be used to easily extract the year from the date

In [ ]:
fn_extract_year(df_input.head(4),column='BIRTHDATE')

### Term Mapping

In the term mapping, the structural mapping `rule_id` is mapped telling us how to map a source term to a destination term, i.e. if the source term is `M` then the output should be `8507`

In [ ]:
f_term_mapping = 'sample_input_data/rules2.csv'
etltool.load_term_mapping(f_term_mapping)
etltool.get_term_mapping_df()

### Run the tool

In [ ]:
etltool.run()

We can finally get the output in a dataframe

In [ ]:
etltool.get_output_df('person')